### PR TITLE
Improvements to buffer pool

### PR DESCRIPTION
**Pull request by lthibault**
- [x] Removes unused buffer pool buckets
- [x] Improves buffer reuse by consuming buffers even if capacity is not aligned to bucket boundary
- [x] Aggressively reuses small-sized buffers by setting min page size to 1kb
- [x] Improves buffer reuse by decreasing total number of buckets
- [x] Small readability improvements (use `range` instead of classic `for` loop)
- [x] Increase max page size to 1Mb

### CODE REVIEW

File: message.go [hunk 1]
```diff
@@ -465,8 +462,9 @@ func (ssa *SingleSegmentArena) Allocate(sz Size, segs map[SegmentID]*Segment) (S
 	if err != nil {
 		return 0, nil, err
 	}
-	buf := make([]byte, len(data), cap(data)+inc)
-	copy(buf, data)
+	buf := bufferpool.Default.Get(cap(data) + inc)
+	buf = buf[:copy(buf, data)]
```
Comment (zenhack) [inline message.go:466]: Shoving all this on one line makes it hard to follow imo; maybe:

```
copied := copy(buf, data)
buf = buf[:copied]
```

File: exp/bufferpool/pool.go [hunk 2]
```diff
@@ -18,33 +19,29 @@ type Pool struct {
...
-	for i := 0; i < len(buf); i++ {
+	for i := range buf {
 		buf[i] = 0
 	}
 
 	capacity := cap(buf)
-	for i := 0; i < nBuckets; i++ {
-		if (1 << i) == capacity {
+	for i := range p.buckets {
+		if (1 << i) >= capacity {
```
Comment (zenhack) [inline exp/bufferpool/pool.go:44]: - Changing this to `>=` is going to result in undersized buffers in the buckets, which is going to break things if someone asks for a buffer bigger than what's returned but <= the pool size, unless I'm missing something.
- Do we need to change `1 << i` to `minSize << i` here as well?

...we should probably add some tests for this package, so we catch this kind of thing.

File: message.go [hunk 3]
```diff
@@ -485,27 +483,8 @@ func (ssa SingleSegmentArena) String() string {
...
-func singleSegment(b []byte) *SingleSegmentArena {
-	return (*SingleSegmentArena)(&b)
-}
-
-var singleSegmentPool = sync.Pool{
-	New: func() any {
-		return singleSegment(nil)
-	},
+	bufferpool.Default.Put(*ssa)
+	(*ssa) = nil
```
Comment (zenhack) [inline message.go:487]: Nit: don't need the parens on the lvalue

File: exp/bufferpool/pool.go [hunk 4]
```diff
@@ -4,7 +4,8 @@ package bufferpool
 import "sync"
 
 const (
-	nBuckets = 20
+	minSize  = 1024
+	nBuckets = 11 // max 1Mb
```
Comment (zenhack) [inline exp/bufferpool/pool.go:8]: Maybe spell out the math behind this a bit more, or (better, if you can figure out how) somehow define it as an expression.

Also, nit: the unit you're looking for is MiB.

Also, I wish Go had static assertions.

**Review by zenhack — CHANGES_REQUESTED** [review]
A couple things in-line

File: exp/bufferpool/pool.go (hunk 4 above)
Comment (lthibault) [reply exp/bufferpool/pool.go:8]: >Also, nit: the unit you're looking for is MiB.

Thanks - for some reason, I can never remember this 😆

File: exp/bufferpool/pool.go (hunk 2 above)
Comment (lthibault) [reply exp/bufferpool/pool.go:44]: Ugh, this is a late-night mistake.  The intent was to allow buffers larger than the bucket's size, but smaller than the _next_ bucket's size.  Obviously this code doesn't do that, though, so I'll fix it.

File: message.go (hunk 1 above)
Comment (lthibault) [reply message.go:466]: This shouldn't have been pushed to this PR.  I've corrected it in #465.

File: message.go (hunk 3 above)
Comment (lthibault) [reply message.go:487]: Ditto. Fixed in #465.

File: exp/bufferpool/pool.go [hunk 5]
```diff
@@ -18,33 +19,32 @@ type Pool struct {
...
+	for i := range buf {
 		buf[i] = 0
 	}
 
 	capacity := cap(buf)
-	for i := 0; i < nBuckets; i++ {
-		if (1 << i) == capacity {
+	for i := range p.buckets {
+		bucket := (minSize << i)
+		next := (minSize<<i + 1)
```
Comment (zenhack) [inline exp/bufferpool/pool.go:45]: Maybe `minSize << (i + 1)`? This might be right, but I can't remember the precedence rules well enough without looking it up, and it would be clearer I think.

Also the parens on the first one are entirely superfluous.

**Review by zenhack — CHANGES_REQUESTED** [review]
One more minor style nit

**Comment by lthibault** [comment]
I ended up re-writing a significant portion of this to allow users to provide their own pool implementations.  I think this can have a very large impact in niche cases (e.g. embedded systems), where you might want to tune the pooling behavior to save memory.  I can also imagine applications that allocate a large volume of very big segments.

I've also added tests around the sensitive parts of the default pool implementation.

File: capability.go [hunk 6]
```diff
@@ -397,10 +397,10 @@ func (c Client) SendCall(ctx context.Context, s Send) (*Answer, ReleaseFunc) {
 
 // SendStreamCall is like SendCall except that:
 //
-// 1. It does not return an answer for the eventual result.
-// 2. If the call returns an error, all future calls on this
-//    client will return the same error (without starting
-//    the method or calling PlaceArgs).
+//  1. It does not return an answer for the eventual result.
```
Comment (lthibault) [inline capability.go:400]: Not sure what happened here.  gofmt, maybe?

**Comment by lthibault** [comment]
Weird, I'm getting [build failures](https://github.com/capnproto/go-capnp/actions/runs/4271716582/jobs/7436377807).  Re-running 🤨

**Comment by lthibault** [comment]
Ok fixed. Ready for review.

File: exp/bufferpool/pool.go [hunk 7]
```diff
@@ -4,49 +4,161 @@ package bufferpool
...
-			item := p.buckets[i].Get()
-			if item == nil {
-				ret = make([]byte, capacity)
-			} else {
-				ret = item.([]byte)
-			}
-			ret = ret[:size]
-			return ret
-		}
+var pool Pool = &BasicPool{}
```
Comment (zenhack) [inline exp/bufferpool/pool.go:11]: Is there a reason for this to no longer be public?

I might name it `defaultPool`; `pool` is a bit too generic. Or keep it as `Default` if it's going to be exported.

File: exp/bufferpool/pool.go [hunk 8]
```diff
@@ -4,49 +4,161 @@ package bufferpool
...
+func Set(p Pool) {
+	if p == nil {
+		panic("Pool(nil)")
 	}
-	return make([]byte, size)
+
+	pool = p
+}
+
+// Get a buffer of with the specified size from the global pool.  The
```
Comment (zenhack) [inline exp/bufferpool/pool.go:31]: Grammar.

File: exp/bufferpool/pool.go [hunk 9]
```diff
@@ -4,49 +4,161 @@ package bufferpool
...
+// resulting buffer's capacity MAY exceed size. The buffer SHOULD NOT
+// be resized beyond its underlying capacity, or it will leak to the
+// garbage collector.
```
Comment (zenhack) [inline exp/bufferpool/pool.go:34]: I might just replace all this with something to the effect of `calls .Get() on the global Pool`, maybe moving some of the details here to that doc comment. Same thing applies to Put()

File: exp/bufferpool/pool.go [hunk 10]
```diff
@@ -4,49 +4,161 @@ package bufferpool
...
+// Put returns the buffer to the global pool.  The slice will be zeroed
+// before returning it to the pool.   Note that only the first len(buf)
+// bytes are zeroed, and not the full backing array.
+//
+// TODO:  should we zero out the entire array?  Users may have resized
+// the slice, so it may contain sensitive data.
+func Put(buf []byte) {
+	for i := range buf {
 		buf[i] = 0
 	}
```
Comment (zenhack) [inline exp/bufferpool/pool.go:48]: Should we be doing this here, or in the implementation of Pool.Put?

File: exp/bufferpool/pool.go [hunk 11]
```diff
@@ -4,49 +4,161 @@ package bufferpool
...
-	capacity := cap(buf)
-	for i := 0; i < nBuckets; i++ {
-		if (1 << i) == capacity {
-			p.buckets[i].Put(buf[:capacity])
+	pool.Put(buf[:cap(buf)])
+}
+
+// Pool is a free list of []byte buffers of heterogeneous sizes.
+// Implementations MUST be thread safe and MUST support efficient
+// paging of buffers with arbitrary capacity.   Most applications
```
Comment (zenhack) [inline exp/bufferpool/pool.go:57]: I'm not exactly sure what paging means in this context.

File: exp/bufferpool/pool.go [hunk 12]
```diff
@@ -4,49 +4,161 @@ package bufferpool
...
+// and Min=1024 (1KiB - ~1MiB buffer sizes).  Most applications
+// will not benefit from tuning these parameters.
+//
+// As a general rule, increasing Min reduces GC latency at the
+// expense of increased memory usage.  Increasing N can reduce
+// GC latency in applications that frequently allocate buffers
+// of size >=1MiB.
+type BasicPool struct {
+	once    sync.Once
+	Min, N  int
```
Comment (zenhack) [inline exp/bufferpool/pool.go:86]: Maybe `MinSize` and `BucketCount`? `N` in particular as a field name feels not descriptive enough.

**Review by zenhack — CHANGES_REQUESTED** [review]
Some comments in-line

File: exp/bufferpool/pool.go (hunk 7 above)
Comment (lthibault) [reply exp/bufferpool/pool.go:11]: Two reasons:

1. avoid it being set to `nil`
2. force calls to go through the top-level `Get` and `Put` methods, since they handle things like zeroing memory, which keeps the interface implementation simple.

Comment (lthibault) [reply exp/bufferpool/pool.go]: >I might name it defaultPool; pool is a bit too generic. 

Done.

File: exp/bufferpool/pool.go (hunk 10 above)
Comment (lthibault) [reply exp/bufferpool/pool.go:48]: I personally like it here, as it keeps the implementation of `Pool` simpler.  It also prevents user-supplied implementations from forgetting to do this.

File: exp/bufferpool/pool.go (hunk 12 above)
Comment (lthibault) [reply exp/bufferpool/pool.go]: Done.

File: exp/bufferpool/pool.go (hunk 11 above)
Comment (lthibault) [reply exp/bufferpool/pool.go:57]: Fixed (dunno what I was thinking).

File: exp/bufferpool/pool.go (hunk 10 above)
Comment (lthibault) [reply exp/bufferpool/pool.go:48]: @zenhack what do you think of the TODO?  Any reason why we're not zeroing out the entire array?  Is it out of concern for performance?

File: exp/bufferpool/pool.go (hunk 9 above)
Comment (lthibault) [reply exp/bufferpool/pool.go]: Done

File: exp/bufferpool/pool.go [hunk 13]
```diff
@@ -4,49 +4,151 @@ package bufferpool
...
+	defaultPool = p
+}
+
+// Get calls .Get() on the global pool, and resizes the resulting buf
+// to size.
+func Get(size int) []byte {
+	return defaultPool.Get(size)[:size]
+}
+
+// Put calls .Put(buff) on the global pool after zeroing the supplied
```
Comment (zenhack) [inline exp/bufferpool/pool.go:37]: I really think we should move the zeroing logic back into Pool.Put, *especially* now that the interface is gone. Same thing with the capacity check.

File: exp/bufferpool/pool.go [hunk 14]
```diff
@@ -4,49 +4,151 @@ package bufferpool
...
+	}
+}
+
+// Pool maintains a list of BucketCount buckets containing buffers
+// of exponentially-increasing size, starting from MinSize. Buffers
+// whose capacity exceeds MinSize << BucketCount are ignored and left
+// to the garbage-collector.
+//
+// The zero-value Pool is ready to use, defaulting to BucketCount=10
+// and MinSize=1024 (1KiB - ~1MiB buffer sizes).  Most applications
```
Comment (zenhack) [inline exp/bufferpool/pool.go:59]: Nit: the tilde is unnecessary; Mi is a power of two prefix, so this is actually exact (the base 10 prefixes don't have the i).

File: exp/bufferpool/pool.go [hunk 15]
```diff
@@ -4,49 +4,151 @@ package bufferpool
...
-	buckets [nBuckets]sync.Pool
+	once                 sync.Once
+	MinSize, BucketCount int
+	buckets              []bucket
 }
 
-// Get a buffer of the given length. Its capacity may be larger than the
-// requested size.
+// Get a buffer of len >= size.  The caller will resize the
+// buffer to size, so implementations need not bother.  The
```
Comment (zenhack) [inline exp/bufferpool/pool.go:73]: The "implementations..." language should probably change now that this is not an interface.

**Review by zenhack — CHANGES_REQUESTED** [review]

File: exp/bufferpool/pool.go (hunk 14 above)
Comment (lthibault) [reply exp/bufferpool/pool.go:59]: Unless my arithmetic is mistaken (totally possible), the slices contained in the largest bucket will have a capacity of at most 1048575 Bytes (.99 MiB).  See `TryPut`:

```
func (b bucket) TryPut(buf []byte) (ok bool) {
	if ok = b.minCap <= cap(buf) && b.capLimit > cap(buf); ok {
		b.pool.Put(buf)
	}

	return
}
```

File: exp/bufferpool/pool.go (hunk 13 above)
Comment (lthibault) [reply exp/bufferpool/pool.go:37]: If we're not splitting the zeroing vs pooling logic, then I see no reason to make the default pool a private variable.  I'll revert back to what we had before and remove the top-level `Get` and `Put` functions.

Comment (lthibault) [reply exp/bufferpool/pool.go]: Done.

File: exp/bufferpool/pool.go (hunk 14 above)
Comment (zenhack) [reply exp/bufferpool/pool.go:21]: Bucket i should have capacity `1024 << i` = `1024 * 2 ^ i` = `2 ^ 10 * 2 * i` = `2 ^ (10 + i)` so with 10 buckets the largest should be `2 ^ 20` = 1MiB. What am I missing?

Comment (zenhack) [reply exp/bufferpool/pool.go:21]: Actually, with 10 buckets the last bucket is `i = 9`, so the limit is actually half a MiB...

File: exp/bufferpool/pool.go [hunk 16]
```diff
@@ -1,34 +1,38 @@
...
+// avoidance of doubt:  calls to Get() with size < MinAlloc return a
+// buffer of len(buf) = size and cap(buf) >= MinAlloc. MinAlloc MUST
+// NOT exceed 1 << BucketCount, or method calls to Pool will panic.
 //
-// The zero-value Pool is ready to use, defaulting to BucketCount=10
-// and MinSize=1024 (1KiB - ~1MiB buffer sizes).  Most applications
-// will not benefit from tuning these parameters.
+// The zero-value Pool is ready to use, defaulting to BucketCount=20
+// and MinAlloc=1024 (max size = ~1MiB).  Most applications will not
+// benefit from tuning these parameters.
```
[the remark this answers is not in the document]
Comment (lthibault) [reply exp/bufferpool/pool.go:26]: Did you see my message on Matrix?  I ended up changing the implementation a bit.
(Apologies for the noise.  The scope of this PR ended up shifting around more than I'd like.  Maybe I should close it and open a fresh one?)

File: exp/bufferpool/pool.go [hunk 17]
```diff
@@ -1,52 +1,135 @@
...
+		for idx = range p.buckets {
+			if 1<<idx >= p.MinAlloc {
+				break
+			}
+		}
+
+		// Set the 'New' function for all buckets lower than 'MinSize'
+		// to the MinSize bucket's 'New' function.  This ensures that
+		// we always allocate a capacity of *at least* MinSize, while
+		// still re-using small buffers.
```
[the remark this answers is not in the document]
Comment (lthibault) [reply exp/bufferpool/pool.go:92]: That would result in undersized buffers being returned to `p.buckets[idx]`.

[the remark this answers is not in the document]
Comment (lthibault) [reply exp/bufferpool/pool.go]: I've reworked this a bit to make it more readable.

[the remark this answers is not in the document]
Comment (zenhack) [reply exp/bufferpool/pool.go]: But if we're returning oversized buffers anyway, shouldn't it be fine?

[the remark this answers is not in the document]
Comment (zenhack) [reply exp/bufferpool/pool.go]: Per discussion on matrix, for posterity: this is so that it's possible to return buffers not allocated with the Pool in the first place, so we'll keep it as-is.

File: exp/bufferpool/pool.go [hunk 18]
```diff
@@ -1,52 +1,117 @@
...
+			return bs[i].Get().([]byte)
+		}
+	}
+
+	return nil
+}
+
+func (bs bucketSlice) Put(buf []byte) {
+	for i := range bs {
+		if 1<<i <= cap(buf) {
```
Comment (lthibault) [inline exp/bufferpool/pool.go:97]: Whoops, this is a bug.  Fixing.

**Review by zenhack — DISMISSED** [review]

**Comment by lthibault** [comment]
Ok, I think this is _finally_ ready to merge.  How do you feel about squashing this branch?

**Comment by zenhack** [comment]
Once we figure out CI, SGTM.
